src/app/forms/custom-sections/required-standard/required-standard.component.ts

```diff
@@ -1,11 +1,135 @@
+import { ResultOfTestService } from '@services/result-of-test/result-of-test.service';
 import { DefaultNullOrEmpty } from '@shared/pipes/default-null-or-empty/default-null-or-empty.pipe';
+import { getRequiredStandardFromTypeAndRef } from '@store/required-standards/selectors/required-standards.selector';
+import { selectRouteParam } from '@store/router/selectors/router.selectors';
+import {
+  createRequiredStandard, removeRequiredStandard, testResultInEdit, toEditOrNotToEdit, updateRequiredStandard,
+} from '@store/test-records';
+import {
+  Subject, distinctUntilChanged,
+  takeUntil, withLatestFrom,
+} from 'rxjs';
 
 @Component({
   selector: 'app-required-standard',
   templateUrl: './required-standard.component.html',
   providers: [DefaultNullOrEmpty],
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
-export class RequiredStandardComponent {
+export class RequiredStandardComponent implements OnInit, OnDestroy {
+  form!: CustomFormGroup;
+  index!: number;
+  requiredStandard?: TestResultRequiredStandard;
+  onDestroy$ = new Subject();
+  isEditing: boolean;
+  amendingRs?: boolean;
+
+  private requiredStandardForm?: CustomFormArray;
+
+  constructor(
+    private store: Store,
+    private activatedRoute: ActivatedRoute,
+    private resultService: ResultOfTestService,
+    private router: Router,
+    private dfs: DynamicFormService,
+    private errorService: GlobalErrorService,
+  ) {
+    this.isEditing = this.activatedRoute.snapshot.data['isEditing'];
+  }
+
+  ngOnInit(): void {
+    const inspectionType = this.store.pipe(select(selectRouteParam('inspectionType')));
+    const rsRefCalculation = this.store.pipe(select(selectRouteParam('ref')));
+    const requiredStandardIndex = this.store.pipe(select(selectRouteParam('requiredStandardIndex')));
+
+    this.store.pipe(select(this.isEditing ? testResultInEdit : toEditOrNotToEdit)).pipe(
+      withLatestFrom(inspectionType, rsRefCalculation, requiredStandardIndex),
+      takeUntil(this.onDestroy$),
+      distinctUntilChanged((prev, curr) => prev[0]?.testTypes[0]?.testResult === curr[0]?.testTypes[0]?.testResult),
+    ).subscribe(([testResult, inspectionTypeValue, rsRefCalculationValue, requiredStandardIndexValue]) => {
+      if (!testResult) this.navigateBack();
+      this.requiredStandardForm = (this.dfs.createForm(RequiredStandardsTpl, testResult) as CustomFormGroup)
+        .get(['testTypes', '0', 'requiredStandards']) as CustomFormArray;
+      if (requiredStandardIndexValue) {
+        this.amendingRs = true;
+        this.index = Number(requiredStandardIndexValue);
+        this.form = this.requiredStandardForm.controls[this.index] as CustomFormGroup;
+        this.requiredStandard = testResult?.testTypes[0]?.requiredStandards?.at(this.index);
+      } else {
+        this.amendingRs = false;
+        this.store.pipe(select(getRequiredStandardFromTypeAndRef(inspectionTypeValue as INSPECTION_TYPE, rsRefCalculationValue ?? '')))
```

Comment:
```suggestion
        this.store.pipe(select(getRequiredStandardFromTypeAndRef(inspectionTypeValue as INSPECTION_TYPE, rsRefCalculationValue ?? '')), takeUntil(this.onDestroy$))
```

Reply:
I can't remember how to fix this, my memory has leaked

Reply:
thanks everyone, here all week